File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def combinationSum(self, nums, target):
        res = []

        def backtrack(start, path, total):
            if total == 0:
                res.append(path[:])
                return
            if total < 0:
                return

            for i in range(start, len(nums)):
                path.append(nums[i])
                backtrack(i, path, total - nums[i])
                path.pop()

        backtrack(0, [], target)
        return res

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        helper(res,new ArrayList<>(),candidates,target,0);
        return res;
    }
    public void helper(List<List<Integer>> res, List<Integer> ans, int[] cand, int target, int idx){
        if(target == 0){
            res.add(new ArrayList<>(ans));
            return;
        }
        if(target < 0 || idx >= cand.length) return;
        helper(res,ans,cand,target,idx+1);
        ans.add(cand[idx]);
        helper(res,ans,cand,target-cand[idx],idx);
        ans.remove(ans.size()-1);
    }
}